Userland/Libraries/LibGfx/ImageFormats/AVIFLoader.cpp

```diff
@@ -0,0 +1,194 @@
+}
+
+static ErrorOr<void> decode_avif_image(AVIFLoadingContext& context)
+{
+    VERIFY(context.state >= AVIFLoadingContext::State::HeaderDecoded);
+
+    avifRGBImage rgb;
+    while (avifDecoderNextImage(context.decoder) == AVIF_RESULT_OK) {
+        auto bitmap_format = context.has_alpha ? BitmapFormat::BGRA8888 : BitmapFormat::BGRx8888;
+        auto bitmap = TRY(Bitmap::create(bitmap_format, context.size.value()));
+
+        avifRGBImageSetDefaults(&rgb, context.decoder->image);
+        rgb.pixels = bitmap->scanline_u8(0);
+        rgb.rowBytes = bitmap->pitch();
+        rgb.format = avifRGBFormat::AVIF_RGB_FORMAT_BGRA;
+
+        avifResult result = avifImageYUVToRGB(context.decoder->image, &rgb);
+        if (result != AVIF_RESULT_OK)
+            return Error::from_string_literal("Conversion from YUV to RGB failed");
+
+        auto duration = context.decoder->imageCount == 1 ? 0 : static_cast<int>(context.decoder->imageTiming.duration * 1000);
+        context.frame_descriptors.append(ImageFrameDescriptor { bitmap, duration });
+
+        context.state = AVIFLoadingContext::BitmapDecoded;
+    }
+
+    return {};
+}
+
+IntSize AVIFImageDecoderPlugin::size()
+{
+    return m_context->size.value();
+}
+
+bool AVIFImageDecoderPlugin::sniff(ReadonlyBytes data)
+{
+    AVIFLoadingContext context;
+    context.data = data;
+
+    auto result = !decode_avif_header(context).is_error();
+
+    // Need to explicitly destroy the decoder since this method is static to prevent leaks.
+    avifDecoderDestroy(context.decoder);
+    context.decoder = nullptr;
+
+    return result;
+}
+
+ErrorOr<NonnullOwnPtr<ImageDecoderPlugin>> AVIFImageDecoderPlugin::create(ReadonlyBytes data)
+{
+    auto context = TRY(try_make<AVIFLoadingContext>());
+    auto plugin = TRY(adopt_nonnull_own_or_enomem(new (nothrow) AVIFImageDecoderPlugin(data, move(context))));
+    TRY(decode_avif_header(*plugin->m_context));
+    return plugin;
+}
+
+bool AVIFImageDecoderPlugin::is_animated()
+{
+    return m_context->image_count > 1;
+}
+
+size_t AVIFImageDecoderPlugin::loop_count()
+{
+    return is_animated() ? m_context->repetition_count : 0;
+}
+
+size_t AVIFImageDecoderPlugin::frame_count()
+{
+    if (!is_animated())
+        return 1;
+    return m_context->image_count;
+}
+
+size_t AVIFImageDecoderPlugin::first_animated_frame_index()
+{
+    return 0;
+}
+
+ErrorOr<ImageFrameDescriptor> AVIFImageDecoderPlugin::frame(size_t index, Optional<IntSize>)
+{
+    if (index >= frame_count())
+        return Error::from_string_literal("AVIFImageDecoderPlugin: Invalid frame index");
+
+    if (m_context->state == AVIFLoadingContext::State::Error)
+        return Error::from_string_literal("AVIFImageDecoderPlugin: Decoding failed");
+
+    if (m_context->state < AVIFLoadingContext::State::BitmapDecoded) {
+        TRY(decode_avif_image(*m_context));
+        m_context->state = AVIFLoadingContext::State::BitmapDecoded;
+    }
+
+    return m_context->frame_descriptors[index];
```

Comment:
Accessing the image without checking that it exists in the `Vector` is a dangerous game

Reply:
The operator appears to call `Vector::at()` which has a verification that this is in bounds. I can add a manual one here, though, too. Is that what you meant?

Reply:
`at()` will make the decoder crash, so yes, please check that manually!